util/pools/resource_pool.go

```diff
@@ -0,0 +1,149 @@
+// license that can be found in the LICENSE file.
+
+// Package pools provides functionality to manage and reuse resources
+// like connections.
+package pools
+
+import (
+	"fmt"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+type Factory func() (Resource, error)
+
+// Resource is an abstraction for all data structures
+// that can be pooling. Thread synchronization between Close()
+// and IsClosed() is the responsibility the caller.
+type Resource interface {
+	Close()
+}
+
+// ResourcePool allows you to use a pool of resources.
+type ResourcePool struct {
+	resources   chan resourceWrapper
+	factory     Factory
+	capacity    *atomic.Int64
+	idleTimeout *atomic.Duration
+	waitCount   *atomic.Int64
+	waitTime    *atomic.Duration
+}
+
+type resourceWrapper struct {
+	resource Resource
+	timeUsed time.Time
+}
+
+// NewResourcePool creates a new ResourcePool pool.
+// capacity is the initial capacity of the pool.
+// maxCap is the maximum capacity.
+// If a resource is unused beyond idleTimeout, it's discarded.
+// An idleTimeout of 0 means that there is no timeout.
+func NewResourcePool(factory Factory, capacity, maxCap int, idleTimeout time.Duration) *ResourcePool {
+	if capacity <= 0 || maxCap <= 0 || capacity > maxCap {
+		panic(fmt.Errorf("invalid/out of range capacity"))
+	}
+	rp := &ResourcePool{
+		resources:   make(chan resourceWrapper, maxCap),
+		factory:     factory,
+		capacity:    atomic.NewInt64(int64(capacity)),
+		idleTimeout: atomic.NewDuration(idleTimeout),
+	}
+	for i := 0; i < capacity; i++ {
+		rp.resources <- resourceWrapper{}
+	}
+	return rp
+}
+
+// Close empties the pool calling Close on all its resources.
+// You can call Close while there are outstanding resources.
+// It waits for all resources to be returned (Put).
+// After a Close, Get are not allowed.
+func (rp *ResourcePool) Close() {
+	// Atomically swap new capacity with old, but only
+	// if old capacity is non-zero.
+	var capacity int
+	for {
+		capacity = int(rp.capacity.Load())
+		if capacity == 0 {
+			return
+		}
+		if rp.capacity.CAS(int64(capacity), int64(0)) {
+			break
+		}
+	}
+
+	for i := 0; i < capacity; i++ {
+		wrapper := <-rp.resources
+		if wrapper.resource != nil {
+			wrapper.resource.Close()
+		}
+	}
+
+	close(rp.resources)
+}
+
+// IsClosed returns whether the ResourcePool is closed.
+func (rp *ResourcePool) IsClosed() (closed bool) {
+	return rp.capacity.Load() == 0
+}
+
+// Get will return the next available resource. If capacity
+// has not been reached, it will create a new one using the factory. Otherwise,
+// it will indefinitely wait till the next resource becomes available.
+func (rp *ResourcePool) Get() (resource Resource, err error) {
+	// Fetch
+	var wrapper resourceWrapper
+	var ok bool
+	wrapper, ok = <-rp.resources
+	if !ok {
+		return nil, fmt.Errorf("ResourcePool is closed")
+	}
+
+	// Unwrap
+	timeout := rp.idleTimeout.Load()
+	if wrapper.resource != nil && timeout > 0 && wrapper.timeUsed.Add(timeout).Sub(time.Now()) < 0 {
```

Comment:
This line recycles resource only if we call `Get` and finally find that it's timeout. Then we close it and immediately recreate a new resource =。=

cc @ngaut @tiancaiamao

Reply:
A pretty way is using Since method.
```go
time.Since(wrapper.timeUsed) > timeout 
```

Reply:
@ngaut I don't talk about code style here, but we don't have a background task to do the actual recycle task. So that the whole recycle alike logic here is a waste - we close the resource and immediately create a new one.

See also [the latest upstream code](https://github.com/vitessio/vitess/blob/896bc7cb6b8c7363657e76239d16a99249addd73/go/pools/resource_pool.go#L170-L192) they start a goroutine for recycling resources.

Reply:
I'm thinking whether we can just use `sync.Pool` which is integrated with gc procedure. The shortcoming can be `Pool`'s `New` cannot return error so that we may need some workaround.

Reply:
Got your point, we can just simply keep update to upstream code.

Reply:
Fine. Let me give it a try. I'm thinking whether we need such a complex solution :P

Reply:
Because the upstream code doesn't intend to be used as a third-party library: it depends on vitess internal `sync2` (which can be replaced with `uber-go/atomic`) and `timer` (which I don't have idea but copy its source code). Vitess itself bundle as a whole application.

I'll create an issue to exploring what's the better way to improve this `ResourcePool` functionality implementation.

Reply:
I don't get the point for adding the 'VITENESS_LICENSE', why it is required after this PR?

Reply:
@bb7133 this PR isn't to be merged. I'll close it now.

You can start a new comment to discussion about LICENSE instead of following this one talking about code.

Basically, pools code is copy from vitess and according to the BSD-3 Clause license requirement we should keep the license file and notice file if any. Our project is quite lack of the awareness about license compliance issues.